### PR TITLE
Use ExceptionDispatchInfo to preserve stack trace

### PR DESCRIPTION
**Pull request by jorgebay**

### CODE REVIEW

File: src/Cassandra.IntegrationTests/Core/ExceptionsTests.cs [hunk 1]
```diff
@@ -15,67 +15,90 @@
...
+        private ITestCluster _testCluster;
+        private ISession _session;
+        private const string KeyspaceNameWriteFailName = "ks_wfail";
+        private string _keyspace;
+        private string _table;
+
+        [OneTimeSetUp]
+        public void TestFixtureSetup()
+        {
+            _testCluster = TestClusterManager.CreateNew(2, new TestClusterOptions
```
Comment (jorgebay) [inline src/Cassandra.IntegrationTests/Core/ExceptionsTests.cs]: These tests were built before scassandra / simulacron integration.
I think we should avoid trying to test it with ccm and do it with simulacron.

**Comment by landim** [comment]
LGTM 👍

File: src/Cassandra.IntegrationTests/TestClusterManagement/SimulacronManager.cs [hunk 2]
```diff
@@ -30,10 +30,7 @@ public static Uri BaseAddress
 
         public void Start()
         {
-            if (_started)
-            {
-                return;
-            }
+            Stop();
```
Comment (jorgebay) [inline src/Cassandra.IntegrationTests/TestClusterManagement/SimulacronManager.cs]: Without going much into detail, its weird to have a `Start()` method that calls a `Stop()` method...

Can we move simulacron fixes and exception test fixes out of this pull request? I think the changes are not directly related.

File: src/Cassandra.IntegrationTests/CommonFixtureSetup.cs [hunk 3]
```diff
@@ -20,15 +20,13 @@ public void SetupTestSuite()
...
             Trace.TraceInformation("Starting Test Run ...");
-            SimulacronManager.Instance.Start();
         }
 
         [OneTimeTearDown]
         public void TearDownTestSuite()
         {
             // this method is executed once after all the fixtures have completed execution
             TestClusterManager.TryRemove();
-            SimulacronManager.Instance.Stop();
```
Comment (jorgebay) [inline src/Cassandra.IntegrationTests/CommonFixtureSetup.cs]: We should reset this change too, along with the rest of `CommonFixtureSetup` changes. I think this is preventing for the appveyor build to finish.

File: src/Cassandra.IntegrationTests/TestClusterManagement/CcmCluster.cs [hunk 4]
```diff
@@ -36,6 +36,10 @@ public void Create(int nodeLength, TestClusterOptions options = null)
             }
             _ccm = new CcmBridge(Name, ClusterIpPrefix);
             _ccm.Create(_version, options.UseSsl);
+            if (options.CassandraYaml != null)
+            {
+                _ccm.UpdateConfig(options.CassandraYaml);
```
Comment (jorgebay) [inline src/Cassandra.IntegrationTests/TestClusterManagement/CcmCluster.cs]: This is not being used, can we skip this change too? I want to avoid unnecessary conflicts with the DSE driver.

File: src/Cassandra.IntegrationTests/TestClusterManagement/CcmBridge.cs [hunk 5]
```diff
@@ -96,6 +96,18 @@ public void Populate(int dc1NodeLength, int dc2NodeLength, bool useVNodes)
             ExecuteCcm(string.Join(" ", parameters));
         }
 
+        public void UpdateConfig(params string[] configs)
```
Comment (jorgebay) [inline src/Cassandra.IntegrationTests/TestClusterManagement/CcmBridge.cs]: Not being used for this changeset.

File: src/Cassandra.IntegrationTests/Core/ExceptionsTests.cs [hunk 6]
```diff
@@ -110,28 +139,29 @@ public void NoHostAvailableException()
...
+                    data_present = true,
+                    delay_in_ms = 0,
+                    message = "read_timeout",
+                    ignore_on_prepare = false
+                }
+            };
+
+            _simulacronCluster.Prime(primeQuery);
+            var ex = Assert.Throws<ReadTimeoutException>(() =>
+                _session.Execute(new SimpleStatement(cql).SetConsistencyLevel(ConsistencyLevel.All)));
```
Comment (jorgebay) [inline src/Cassandra.IntegrationTests/Core/ExceptionsTests.cs]: As we are dealing with errors, I think it would be best to create a new `Cluster` instance each time and dispose them at the end.

**Comment by jorgebay** [comment]
Besides [the comment from above](https://github.com/datastax/csharp-driver/pull/340#discussion_r145064412), can you rebase it and fix the conflicts on `src/Cassandra.IntegrationTests/Core/ExceptionsTests.cs`?

**Comment by landim** [comment]
Rebased and added `cluster.Shutdown()` at test TearDown

**Comment by jorgebay** [comment]
I've pushed a fix for `PreserveStackTraceTest` under .NET Core, that was causing a failure: https://ci.appveyor.com/project/DataStax/csharp-driver/build/CSHARP-598-1440/job/yo6uddhgbpipgy6c

ExceptionTest changes to use Simulacron look great!

I'll merge it as soon as CI builds finish.
